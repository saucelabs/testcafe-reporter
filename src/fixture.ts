import { TestRun, Test } from '@saucelabs/sauce-json-reporter';

export type Asset = { name: string; localPath: string };
export type Assets = Array<Asset>;

/**
 * Provides a facade to help associate a sauce-json-reporter/TestRun with additional TestCafe specific
 * execution metadata (e.g. browser under test, generated assets).
 */
export class BrowserTestRun {
  userAgent: string;
  testRun: TestRun;
  assets: Assets;

  #browser: string;
  platform: string;

  constructor(userAgent: string) {
    this.userAgent = userAgent;
    this.testRun = new TestRun();
    (this.testRun.metadata as { userAgent: string })['userAgent'] = userAgent;
    this.assets = [];

    // NOTE: example userAgents:
    // * Chrome 126.0.0.0 / Sonoma 14
    // * Chrome 126.0.0.0 / Windows 10 (https://app.saucelabs.com/tests/000aa4ffc86d40bdbeebfcf165dab402)
    const matches = userAgent.match(/^([\w .]+)(?:\/([\w .]+))?/);

    if (matches) {
      this.#browser = matches[1]?.trim() ?? '';
      this.platform = matches[2]?.trim() ?? '';
    } else {
      this.#browser = '';
      this.platform = '';
    }
  }

  /**
   * Returns the job id for the remotely executed test run.
   *
   * Job link is provided in the user agent by the saucelabs browser provider.
   */
  get jobId() {
    // NOTE: example match:
    // * Chrome 126.0.0.0 / Windows 10 (https://app.saucelabs.com/tests/000aa4ffc86d40bdbeebfcf165dab402)
    const matches = this.userAgent.match(
      /https:\/\/.*saucelabs\.com\/tests\/(\w+)/,
    );
    if (!matches || matches.length < 2) {
      return null;
    }
    return matches[1];
  }

  get browserName() {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const [name = 'unknown', _] = this.#browser.split(' ');
    return name;
  }

  get browserVersion() {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const [_, version = 'unknown'] = this.#browser.split(' ');
    return version;
  }

  get browser() {
    return this.#browser;
  }

  addTest(fixturePath: string, fixtureName: string, test: Test) {
    this.testRun.withSuite(fixturePath).withSuite(fixtureName).addTest(test);
    this.testRun.computeStatus();
  }

  addAssets(assets: Assets) {
    this.assets.push(...assets);
  }
}

export class Fixture {
  name: string;
  path: string;
  meta: object;
  browserTestRuns: Map<string, BrowserTestRun>;
  startTime?: Date;
  endTime?: Date;

  constructor(name: string, path: string, meta: object) {
    this.name = name;
    this.path = path;
    this.meta = meta;
    this.browserTestRuns = new Map();
  }

  addTestWithAssets(
    userAgent: string,
    test: Test,
    screenshotAssets: Assets,
    videoAssets: Assets,
  ) {
    if (!this.browserTestRuns.has(userAgent)) {
      this.browserTestRuns.set(userAgent, new BrowserTestRun(userAgent));
    }

    const tr = this.browserTestRuns.get(userAgent) as BrowserTestRun;

    for (const a of screenshotAssets) {
      test.attach({
        path: a.name,
        name: a.name,
        contentType: 'image/png',
      });
    }
    for (const a of videoAssets) {
      test.attach({
        path: a.name,
        name: a.name,
        contentType: 'video/mp4',
      });
    }

    tr.addTest(this.path, this.name, test);
    tr.addAssets([...screenshotAssets, ...videoAssets]);
  }

  get testRuns() {
    return [...this.browserTestRuns.values()].map((bc) => bc.testRun);
  }

  get localBrowserTestRuns() {
    return [...this.browserTestRuns.values()].filter(
      (run) => run.jobId === null,
    );
  }

  get remoteBrowserTestRuns() {
    return [...this.browserTestRuns.values()].filter(
      (run) => run.jobId !== null,
    );
  }
}
