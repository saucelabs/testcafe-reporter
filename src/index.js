import { TestRun } from '@saucelabs/sauce-json-reporter';

const path = require('path');
const fs = require('fs');

const { SauceJsonReporter } = require('./json-reporter');
const { JobReporter } = require('./job-reporter');

/**
 * @typedef {import("./fixture").Fixture} Fixture
 */

module.exports = function () {
  return {
    noColors: !!process.env.SAUCE_NO_COLORS || !!process.env.SAUCE_VM,
    sauceJsonReporter: SauceJsonReporter.newReporter(),

    sauceReportJsonPath:
      process.env.SAUCE_REPORT_JSON_PATH || './sauce-test-report.json',
    disableUpload: process.env.SAUCE_DISABLE_UPLOAD !== undefined,

    // JobReporter
    indentWidth: 2,
    specPath: '',
    relSpecPath: '',
    afterErrorList: false,
    startTime: null,
    startTimes: new Map(),

    // TestCafe Hooks
    reportTaskStart: async function (
      startTime,
      userAgents,
      testCount,
      testStructure,
      properties,
    ) {
      this.sauceJsonReporter.reportTaskStart(startTime, userAgents, testCount);

      if (this.disableUpload) {
        return;
      }

      this.reporter = new JobReporter(this, properties.configuration.sauce);
      this.startTime = startTime;
      this.testCount = testCount;
      this.taskStartConsole(startTime, userAgents, testCount);
    },

    reportFixtureStart: async function (name, specPath, meta) {
      this.sauceJsonReporter.reportFixtureStart(name, specPath, meta);

      if (this.disableUpload) {
        return;
      }

      if (this.specPath && this.specPath !== specPath) {
        // End of currently running spec
        const completedFixture = this.sauceJsonReporter.fixtures.find(
          (f) => f.path === path.relative(process.cwd(), this.specPath),
        );
        await this.reportFixture(completedFixture);
      }

      this.specPath = specPath;
      this.relSpecPath = path.relative(process.cwd(), this.specPath);
      this.specStartConsole(
        this.relSpecPath,
        this.sauceJsonReporter.fixtures.length,
      );

      this.fixtureStartConsole(name, specPath, meta);
    },

    reportTestStart: async function (name, meta, testStartInfo) {
      this.sauceJsonReporter.reportTestStart(name, meta, testStartInfo);

      if (this.disableUpload) {
        return;
      }

      this.startTimes.set(testStartInfo.testId, new Date());
    },

    reportTestDone: async function (name, testRunInfo, meta) {
      this.sauceJsonReporter.reportTestDone(name, testRunInfo, meta);

      if (this.disableUpload) {
        return;
      }

      this.testDoneConsole(name, testRunInfo, meta);
    },

    reportTaskDone: async function (endTime, passed, warnings, result) {
      this.sauceJsonReporter.reportTaskDone(endTime, passed, warnings, result);
      const mergedTestRun = this.sauceJsonReporter.mergeTestRuns();

      fs.writeFileSync(this.sauceReportJsonPath, mergedTestRun.stringify());

      const remoteTestRuns = this.sauceJsonReporter.remoteTestRuns();
      const tasks = [];
      for (const [jobId, runs] of remoteTestRuns) {
        const p = async () => {
          const merged = new TestRun();

          // The user agent is the same for all runs of a given job.
          merged.metadata['userAgent'] = runs.find(
            (run) => run.metadata['userAgent'],
          )?.metadata['userAgent'];

          runs.forEach((run) => {
            for (const suite of run.suites) {
              suite.metadata = run.metadata;
              merged.addSuite(suite);
            }
          });
          merged.computeStatus();

          await this.reporter.attachTestRun(jobId, merged);
        };
        tasks.push(p());
      }

      await Promise.allSettled(tasks);

      if (this.disableUpload) {
        return;
      }

      await this.reportFixture(this.sauceJsonReporter.currentFixture);

      this.taskDoneConsole(endTime, passed, warnings);
    },

    // Extraneous funcs - Used by JobReporter
    taskStartConsole(startTime, userAgents, testCount) {
      this.setIndent(this.indentWidth)
        .newline()
        .useWordWrap(true)
        .write(
          this.chalk.bold('Running tests in:'),
          startTime,
          userAgents,
          testCount,
        )
        .newline();

      userAgents.forEach((ua) => {
        this.write(
          `- ${this.chalk.blue(ua)}`,
          startTime,
          userAgents,
          testCount,
        ).newline();
      });
    },

    /**
     * @param {Fixture} fixture
     */
    async reportFixture(fixture) {
      if (!this.reporter.isAccountSet()) {
        return;
      }

      const browserTestRuns = fixture.localBrowserTestRuns;
      if (browserTestRuns.length === 0) {
        return;
      }

      this.setIndent(this.indentWidth * 3)
        .newline()
        .write(this.chalk.bold.underline('Sauce Labs Test Report'))
        .newline();

      const reportTasks = [];
      for (const browserTestRun of browserTestRuns) {
        const task = new Promise((resolve, reject) => {
          (async () => {
            const session = {
              name: fixture.path,
              startTime: fixture.startTime,
              endTime: new Date(),
              testRun: browserTestRun.testRun,
              browserName: browserTestRun.browserName,
              browserVersion: browserTestRun.browserVersion,
              platformName: browserTestRun.platform,
              assets: browserTestRun.assets,
              userAgent: browserTestRun.userAgent,
            };
            try {
              const job = await this.reporter.reportSession(session);
              this.setIndent(this.indentWidth * 4)
                .write(
                  `* ${browserTestRun.browser}: ${this.chalk.blue.underline(
                    job.url,
                  )}`,
                )
                .newline();

              if (job.assets.length > 0) {
                this.write(`  ${this.chalk.bold('Assets:')}`).newline();
              }
              for (const asset of job.assets) {
                this.write(`    - ${this.chalk.blue.underline(asset)}`);
              }

              await this.reporter.reportTestRun(
                fixture,
                browserTestRun,
                job.id,
              );
              resolve();
            } catch (e) {
              reject(e);
            }
          })();
        });
        reportTasks.push(task);
      }
      await Promise.allSettled(reportTasks);
      this.newline();
    },

    specStartConsole(relSpecPath, index) {
      this.newline()
        .setIndent(this.indentWidth)
        .useWordWrap(true)
        .write(`${index}) ${this.chalk.underline(relSpecPath)}`)
        .newline();
    },

    fixtureStartConsole(name, specPath, meta) {
      this.setIndent(this.indentWidth).useWordWrap(true);

      if (this.afterErrorList) {
        this.afterErrorList = false;
      } else {
        this.newline();
      }

      this.write(name, specPath, meta).newline();
    },

    testDoneConsole(name, testRunInfo, meta) {
      const hasErr = !!testRunInfo.errs.length;
      let symbol = null;
      let nameStyle = null;

      if (testRunInfo.skipped) {
        this.skipped++;

        symbol = this.chalk.cyan('-');
        nameStyle = this.chalk.cyan;
      } else if (hasErr) {
        symbol = this.chalk.red.bold(this.symbols.err);
        nameStyle = this.chalk.red.bold;
      } else {
        symbol = this.chalk.green(this.symbols.ok);
        nameStyle = this.chalk.grey;
      }
      let title = `${symbol} ${nameStyle(name)} (${testRunInfo.durationMs}ms)`;

      this.setIndent(this.indentWidth).useWordWrap(true);

      if (testRunInfo.unstable) {
        title += this.chalk.yellow(' (unstable)');
      }

      if (testRunInfo.screenshotPath) {
        title += ` (screenshots: ${this.chalk.underline.grey(
          testRunInfo.screenshotPath,
        )})`;
      }

      this.write(title, name, testRunInfo, meta);

      this._renderReportData(testRunInfo.reportData, name, testRunInfo, meta);

      if (hasErr) {
        this._renderErrors(testRunInfo.errs, name, testRunInfo, meta);
      }

      this.afterErrorList = hasErr;

      this.newline();
    },

    taskDoneConsole(endTime, passed, warnings) {
      const durationMs = endTime - this.startTime;
      const durationStr = this.moment
        .duration(durationMs)
        .format('h[h] mm[m] ss[s]');
      let footer =
        passed === this.testCount
          ? this.chalk.bold.green(`${this.testCount} passed`)
          : this.chalk.bold.red(
              `${this.testCount - passed}/${this.testCount} failed`,
            );

      footer += this.chalk.grey(` (${durationStr})`);

      if (!this.afterErrorList) {
        this.newline();
      }

      this.setIndent(this.indentWidth)
        .useWordWrap(true)
        .write(footer)
        .newline();

      if (this.skipped > 0) {
        this.write(this.chalk.cyan(`${this.skipped} skipped`)).newline();
      }

      if (warnings.length) {
        this._renderWarnings(warnings);
      }
    },

    _renderReportData(reportData, browsers, name, testRunInfo, meta) {
      if (!reportData) return;

      if (!Object.values(reportData).some((data) => data.length)) return;

      const renderBrowserName = browsers.length > 1;
      const dataIndent = browsers.length > 1 ? 3 : 2;

      this.newline().setIndent(this.indentWidth).write('Report data:');

      browsers.forEach(({ testRunId, prettyUserAgent }) => {
        const browserReportData = reportData[testRunId];

        if (!browserReportData) return;

        if (renderBrowserName) {
          this.setIndent(this.indentWidth * 2)
            .newline()
            .write(prettyUserAgent, name, testRunInfo, meta);
        }

        browserReportData.forEach((data) => {
          this.setIndent(this.indentWidth * dataIndent)
            .newline()
            .write(`- ${data}`, name, testRunInfo, meta);
        });
      });
    },

    _renderErrors(errs, name, testRunInfo, meta) {
      this.setIndent(this.indentWidth * 3).newline();

      errs.forEach((err, idx) => {
        const prefix = this.chalk.red(`${idx + 1}) `);

        this.newline()
          .write(this.formatError(err, prefix), name, testRunInfo, meta)
          .newline()
          .newline();
      });
    },

    _renderWarnings(warnings) {
      this.newline()
        .setIndent(this.indentWidth * 4)
        .write(this.chalk.bold.yellow(`Warnings (${warnings.length}):`))
        .newline();

      warnings.forEach((msg) => {
        this.setIndent(this.indentWidth * 4)
          .write(this.chalk.bold.yellow('--'))
          .newline()
          .setIndent(this.indentWidth * 5)
          .write(msg)
          .newline();
      });
    },

    log(msg) {
      this.write(msg).newline();
    },

    error(msg) {
      this.newline().write(this.chalk.red(msg)).newline();
    },
  };
};
