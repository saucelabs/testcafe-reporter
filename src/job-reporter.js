const crypto = require('crypto');
const fs = require('fs');
const path = require('path');
const stream = require('stream');
const { Status, Test } = require('@saucelabs/sauce-json-reporter');
const { TestComposer } = require('@saucelabs/testcomposer');

const { TestRuns: TestRunsAPI, Jobs: JobsAPI } = require('./api');
const { CI } = require('./ci');

const assetsURLMap = new Map([
  ['us-west-1', 'https://assets.saucelabs.com'],
  ['eu-central-1', 'https://assets.eu-central-1.saucelabs.com'],
  ['staging', 'https://assets.staging.saucelabs.net'],
]);

class JobReporter {
  constructor(logger = console, opts = {}) {
    this.log = logger;

    let reporterVersion = 'unknown';
    try {
      const packageData = JSON.parse(
        fs.readFileSync(path.join(__dirname, '..', 'package.json'), 'utf-8'),
      );
      reporterVersion = packageData.version;
    } catch (e) {
      /* empty */
    }

    this.username = opts.username || process.env.SAUCE_USERNAME;
    this.accessKey = opts.accessKey || process.env.SAUCE_ACCESS_KEY;
    this.build = opts.build || randomBuildID();
    this.tags = opts.tags;
    this.region = opts.region || 'us-west-1';
    if (opts.cleanupArtifactPath && fs.existsSync(opts.cleanupArtifactPath)) {
      fs.rmSync(opts.cleanupArtifactPath, { recursive: true, force: true });
    }

    const userAgent = `testcafe-reporter/${reporterVersion}`;
    this.testComposer = new TestComposer({
      region: this.region,
      username: this.username,
      accessKey: this.accessKey,
      headers: { 'User-Agent': userAgent },
    });

    this.testRunsAPI = new TestRunsAPI({
      region: this.region,
      username: this.username,
      accessKey: this.accessKey,
      headers: {
        'User-Agent': userAgent,
      },
    });

    this.jobsAPI = new JobsAPI({
      region: this.region,
      username: this.username,
      accessKey: this.accessKey,
    });
  }

  isAccountSet() {
    return this.username && this.accessKey;
  }

  /**
   * Reports a fixture from ./json-reporter to the test-runs api.
   * @param {?} fixture - A Fixture object from the ./json-reporter package
   * @param {?} browserTestRun - A BrowserTestRun object from the ./json-reporter package
   * @param {string} jobId
   */
  async reportTestRun(fixture, browserTestRun, jobId) {
    const baseRun = {
      start_time: fixture.startTime.toISOString(),
      end_time:
        (fixture.endTime && fixture.endTime.toISOString()) ||
        new Date().toISOString(),
      path_name: fixture.path,
      platform: 'other',
      type: 'web',
      framework: 'testcafe',
      sauce_job: {
        id: jobId,
      },
      tags: this.tags || [],
      build_name: this.build,
      ci: {
        ref_name: CI.refName,
        branch: CI.refName,
        repository: CI.repo,
        commit_sha: CI.sha,
      },
    };
    const testRun = browserTestRun.testRun;
    // NOTE: TestRuns for TestCafe will have a single root suite that represents
    // the spec. Since the spec path is a separate field in the Insights TestRun,
    // we can ignore it when flattening the tests.
    const tests = this.flattenTests(testRun.suites[0].suites);
    const reqs = tests.map((test) => {
      const req = {
        ...baseRun,
        name: test.name,
        duration: test.duration,
        browser: browserTestRun.browser,
        os: browserTestRun.platform,
        status: test.status,
      };
      if (test.status === Status.Failed) {
        req.errors = [
          {
            message: test.output,
            path: fixture.path,
          },
        ];
      }
      return req;
    });
    await this.testRunsAPI.create(reqs);
  }

  /**
   * Recurses through suites and returns a flattened list of tests.
   * @param {Suite[]} suites
   * @param {string[]} names
   * @returns {Test[]}
   */
  flattenTests(suites, names = []) {
    let tests = [];
    suites.forEach((suite) => {
      tests = tests.concat(
        this.flattenTests(suite.suites, [...names, suite.name]),
      );

      suite.tests.forEach((test) => {
        tests.push(
          new Test(
            [...names, suite.name, test.name].join(' - '),
            test.status,
            test.duration,
            test.output,
            test.startTime,
            test.attachments,
            test.metadata,
            test.code,
            test.videoTimestamp,
          ),
        );
      });
    });
    return tests;
  }

  /**
   * Attaches a test run to an existing Sauce Labs job.
   *
   * @param jobId {string}
   * @param testRun {import('@saucelabs/sauce-json-reporter').TestRun}
   * @returns {Promise<void>}
   */
  async attachTestRun(jobId, testRun) {
    const reportReadable = new stream.Readable();
    reportReadable.push(testRun.stringify());
    reportReadable.push(null);

    try {
      const resp = await this.testComposer.uploadAssets(jobId, [
        {
          filename: 'sauce-test-report.json',
          data: reportReadable,
        },
        this.createConsoleLog(testRun, testRun.metadata['userAgent']),
      ]);

      if (resp.errors) {
        for (const err of resp.errors) {
          console.error('Failed to upload asset:', err);
        }
      }
    } catch (e) {
      console.error('Failed to upload test result:', e.message);
    }
  }

  /**
   * @param jobId {string}
   * @param passed {boolean}
   */
  async updateJobStatus(jobId, passed) {
    try {
      await this.jobsAPI.updateStatus(jobId, passed);
    } catch (e) {
      console.error('Failed to update job status:', e.message);
    }
  }

  async reportSession(session) {
    const job = await this.testComposer.createReport({
      name: session.name,
      startTime: session.startTime.toISOString(),
      endTime: session.endTime.toISOString(),
      framework: 'testcafe',
      frameworkVersion: '0.0.0', // TODO https://github.com/DevExpress/testcafe/issues/6591
      passed: session.testRun.computeStatus() === Status.Passed,
      build: this.build,
      tags: this.tags,
      browserName: session.browserName,
      browserVersion: session.browserVersion,
      platformName: session.platformName,
    });
    job.assets = [];
    const baseURL = assetsURLMap.get(this.region);

    const assets = session.assets.map((a) => {
      job.assets.push(`${baseURL}/jobs/${job.id}/${a.name}`);

      return {
        filename: a.name,
        data: fs.createReadStream(a.localPath),
      };
    });

    const reportReadable = new stream.Readable();
    reportReadable.push(session.testRun.stringify());
    reportReadable.push(null);
    assets.push({
      filename: 'sauce-test-report.json',
      data: reportReadable,
    });

    assets.push(this.createConsoleLog(session.testRun, session.userAgent));

    await this.testComposer.uploadAssets(job.id, assets).then(
      (resp) => {
        if (resp.errors) {
          for (const err of resp.errors) {
            console.error('Failed to upload asset:', err);
          }
        }
      },
      (e) => console.error('Failed to upload assets:', e.message),
    );

    return job;
  }

  logSuite(suite, depth = 0) {
    const indent = '    '.repeat(depth);
    const resultIcon = suite.status === Status.Passed ? '✓' : '✖';
    let log = `${indent}${resultIcon} ${suite.name}\n`;
    for (const t of suite.tests) {
      log += this.logTest(t, depth + 1);
    }

    for (const s of suite.suites) {
      log += this.logSuite(s, depth + 1);
    }

    return log;
  }

  logTest(test, depth = 0) {
    const indent = '    '.repeat(depth);
    const resultIcon = test.status === Status.Passed ? '✓' : '✖';
    const name = test.name;
    const duration = test.duration;

    let error = '';
    if (test.status !== Status.Passed && test.output !== '') {
      error = `\n\n${indent}\n${test.output}\n\n`;
    }

    return `${indent}${resultIcon} ${name} (${duration}ms)${error}\n`;
  }

  createConsoleLog(testRun, userAgent) {
    let log = `Running tests in: ${userAgent}\n\n\n`;

    log += 'Results:\n\n';
    for (const s of testRun.suites) {
      log += this.logSuite(s, 1);
      log += '\n';
    }

    const r = new stream.Readable();
    r.push(log);
    r.push(null);

    return {
      filename: 'console.log',
      data: r,
    };
  }
}

function randomBuildID() {
  return crypto.randomBytes(6).readUIntLE(0, 6).toString(36);
}

module.exports = { JobReporter };
