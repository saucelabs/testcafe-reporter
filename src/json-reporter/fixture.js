/* eslint-env node */
// @ts-check
// eslint-disable-next-line no-unused-vars
const { TestRun, Test } = require('@saucelabs/sauce-json-reporter');

/**
* Provides a facade to help associate a sauce-json-reporter/TestRun with additional TestCafe specific
* execution metadata (e.g. browser under test, generated assets).
*/
class BrowserTestRun {
    /**
     * @param {string} userAgent
     */
    constructor(userAgent) {
        this.userAgent = userAgent;
        this.testRun = new TestRun();
        // @ts-ignore
        this.testRun.metadata['userAgent'] = userAgent;
        /**
         * @type {Array<{ name: string, localPath: string }>}
         */
        this.assets = [];

        const [ browser, platform ] = userAgent.split('/').map((ua) => ua.trim());

        this._browser = browser;
        this.platform = platform;
    }

    get browserName() {
        // eslint-disable-next-line no-unused-vars
        const [ name = 'unknown',  _ ] = this._browser.split(' ');

        return name;
    }

    get browserVersion() {
        // eslint-disable-next-line no-unused-vars
        const [ _, version = 'unknown' ] = this._browser.split(' ');

        return version;
    }

    get browser() {
        return this._browser;
    }

    /**
     * @param {string} fixturePath
     * @param {string} fixtureName
     * @param {Test} test
     */
    addTest(fixturePath, fixtureName, test) {
        this.testRun.withSuite(fixturePath).withSuite(fixtureName).addTest(test);
        this.testRun.computeStatus();
    }

    /**
     * @param {Array<{ name: string, localPath: string }>} assets
     */
    addAssets(assets) {
        this.assets.push(...assets);
    }
}

class Fixture {
    /**
     * @param {string} name - The fixture name
     * @param {string} path - The path to the file defining the fixture
     * @param {object} meta - Metadata about the fixture
     */
    constructor(name, path, meta) {
        this.name = name;
        this.path = path;
        this.meta = meta;
        /**
         * @type Map<string, BrowserTestRun>
         */
        this.browserTestRuns = new Map();

        /**
         * @type Date | null
         */
        this.startTime = null;
        /**
         * @type Date | null
         */
        this.endTime = null;
    }

    /**
     * @param {string} userAgent
     * @param {Test} test
     * @param {Array<{ name: string, localPath: string }>} screenshotAssets
     * @param {Array<{ name: string, localPath: string }>} videoAssets
     */
    addTestWithAssets(userAgent, test, screenshotAssets, videoAssets) {
        if (!this.browserTestRuns.has(userAgent)) {
            this.browserTestRuns.set(userAgent, new BrowserTestRun(userAgent));
        }

        const tr = this.browserTestRuns.get(userAgent);

        for (const a of screenshotAssets) {
            test.attach({
                path: a.name,
                name: a.name,
                contentType: 'image/png',
            });
        }
        for (const a of videoAssets) {
            test.attach({
                path: a.name,
                name: a.name,
                contentType: 'video/mp4',
            });
        }

        // @ts-ignore
        tr.addTest(this.path, this.name, test);
        // @ts-ignore
        tr.addAssets([...screenshotAssets, ...videoAssets]);
    }

    collectTestRuns() {
        return [...this.browserTestRuns.values()].map((bc) => bc.testRun);
    }
}

module.exports = {
    BrowserTestRun,
    Fixture,
};
