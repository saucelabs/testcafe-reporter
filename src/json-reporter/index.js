const { SauceJsonReporter } = require('./reporter');

module.exports = function() {
    return {
        noColors: true,

        reporter: SauceJsonReporter.newReporter(),

        reportTaskStart (startTime, userAgents, testCount) {
            this.reporter.reportTaskStart(startTime, userAgents, testCount);
        },

        reportFixtureStart (name, path, meta) {
            this.reporter.reportFixtureStart(name, path, meta);
        },

        reportTestStart (name, meta, testStartInfo) {
            this.reporter.reportTestStart(name, meta, testStartInfo);
        },

        reportTestDone (name, testRunInfo, meta) {
            this.reporter.reportTestDone(name, testRunInfo, meta);
        },

        reportTaskDone (endTime, passed, warnings, result) {
            this.reporter.reportTaskDone(endTime, passed, warnings, result);

            const mergedTestRun = this.reporter.mergeTestRuns();
            this.write(mergedTestRun.stringify());
        }
    };
};
