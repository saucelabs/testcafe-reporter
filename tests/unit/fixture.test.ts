import { Status, Test } from '@saucelabs/sauce-json-reporter';
import { Assets, BrowserTestRun, Fixture } from '../../src/fixture';

describe('BrowserTestRun', () => {
  [
    ['Chrome 97.0.4692.71 / macOS 10.15.7', 'Chrome', '97.0.4692.71'],
    ['Firefox 97 / macOS 10.15.7', 'Firefox', '97'],
    ['Firefox 97', 'Firefox', '97'],
    ['Firefox', 'Firefox', 'unknown'],
    [
      'Chrome 126.0.0.0 / Windows 10 (https://app.saucelabs.com/tests/000aa4ffc86d40bdbeebfcf165dab402)',
      'Chrome',
      '126.0.0.0',
    ],
  ].forEach(([userAgent, expectedBrowserName, expectedBrowserVersion]) => {
    test(`can parse browser from userAgent (${userAgent})`, async () => {
      const sut = new BrowserTestRun(userAgent);

      expect(sut.browserName).toBe(expectedBrowserName);
      expect(sut.browserVersion).toBe(expectedBrowserVersion);
    });
  });

  [
    ['Chrome 97.0.4692.71 / macOS 10.15.7', 'macOS 10.15.7'],
    ['Firefox 97 / macOS', 'macOS'],
    [
      'Chrome 126.0.0.0 / Windows 10 (https://app.saucelabs.com/tests/000aa4ffc86d40bdbeebfcf165dab402)',
      'Windows 10',
    ],
  ].forEach(([userAgent, expectedPlatform]) => {
    test(`can parse platform from userAgent (${userAgent})`, async () => {
      const sut = new BrowserTestRun(userAgent);

      expect(sut.platform).toBe(expectedPlatform);
    });
  });

  [
    [
      'Chrome 126.0.0.0 / Windows 10 (https://app.saucelabs.com/tests/000aa4ffc86d40bdbeebfcf165dab402)',
      '000aa4ffc86d40bdbeebfcf165dab402',
    ],
    [
      'Chrome 126.0.0.0 / Windows 10 (https://app.eu-central-1.saucelabs.com/tests/000aa4ffc86d40bdbeebfcf165dab402)',
      '000aa4ffc86d40bdbeebfcf165dab402',
    ],
  ].forEach(([userAgent, expectedJobId]) => {
    test(`can parse jobId from userAgent (${userAgent})`, async () => {
      const sut = new BrowserTestRun(userAgent);

      expect(sut.jobId).toBe(expectedJobId);
    });
  });
});

describe('Fixture', () => {
  test('adding a test adds it to the correct suite', async () => {
    const fixtureName = 'test fixture';
    const fixturePath = 'path/to/some/test.spec.js';
    const fixtureMeta = {};
    const userAgent = 'Chrome 99 /Â macOS 10';
    const screenshotAssets: Assets = [];
    const videoAssets: Assets = [];
    const artifacts: Assets = [];
    const test = new Test(
      'test',
      Status.Passed,
      123,
      '',
      new Date('August 15 2021 00:00:00 UTC'),
    );
    const sut = new Fixture(fixtureName, fixturePath, fixtureMeta);

    sut.addTestWithAssets(
      userAgent,
      test,
      screenshotAssets,
      videoAssets,
      artifacts,
    );

    expect(sut.testRuns).toMatchSnapshot();
  });
});
