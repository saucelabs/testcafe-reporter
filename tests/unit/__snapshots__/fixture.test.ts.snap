// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fixture adding a test adds it to the correct suite 1`] = `
[
  TestRun {
    "attachments": [],
    "metadata": {
      "userAgent": "Chrome 99 /Â macOS 10",
    },
    "status": "passed",
    "suites": [
      Suite {
        "attachments": [],
        "metadata": {},
        "name": "path/to/some/test.spec.js",
        "status": "passed",
        "suites": [
          Suite {
            "attachments": [],
            "metadata": {},
            "name": "test fixture",
            "status": "passed",
            "suites": [],
            "tests": [
              Test {
                "attachments": [],
                "code": undefined,
                "duration": 123,
                "metadata": {},
                "name": "test",
                "output": "",
                "startTime": 2021-08-15T00:00:00.000Z,
                "status": "passed",
                "videoTimestamp": undefined,
              },
            ],
          },
        ],
        "tests": [],
      },
    ],
  },
]
`;
